README.md
# aioinflux

Asynchronous Python client for InfluxDB. 
Built on top of [`aiohttp`](https://github.com/aio-libs/aiohttp) and 
[`asyncio`](https://docs.python.org/3/library/asyncio.html).

InfluxDB is an open-source distributed time series database.
Find more about InfluxDB at http://influxdata.com/


## Installation

Aioinflux is not yet listed on PyPI. Install directly from sources:

    # Latest stable version
    $ pip install git+https://github.com/plugaai/aioinflux
    
    # Latest development commit
    $ pip install git+https://github.com/plugaai/aioinflux@dev


### Dependencies

Aioinflux supports Python 3.6+ **ONLY**. For older Python versions please use the 
[official Python client](https://github.com/influxdata/influxdb-python) 

Third-party library dependencies are: [`aiohttp`](https://github.com/aio-libs/aiohttp) 
for all HTTP request handling and [`pandas`](https://github.com/pandas-dev/pandas) for 
`DataFrame` reading/writing support.


## Usage

### TL;DR:

This sums most of what you can do with `aioinflux`:

```python
import asyncio
from aioinflux import AsyncInfluxDBClient

point = dict(time='2009-11-10T23:00:00Z',
             measurement='cpu_load_short',
             tags={'host': 'server01',
                   'region': 'us-west'},
             fields={'value': 0.64})

client = AsyncInfluxDBClient(db='testdb')

coros = [client.create_database(db='testdb'),
         client.write(point),
         client.query('SELECT value FROM cpu_load_short')]
         
loop = asyncio.get_event_loop() 
results = loop.run_until_complete(asyncio.gather(*coros))
for result in results:
    print(result)
```

### Client modes

Despite its name, `AsyncInfluxDBClient` can also run in sync/blocking modes. 
Avialable modes are: `async` (default), `blocking` and `dataframe`.

Example using `blocking` mode:

```python
client = AsyncInfluxDBClient(db='testdb', mode='blocking')
client.ping()
client.write(point)
client.query('SELECT value FROM cpu_load_short')
```

See [Retrieving DataFrames](#retrieving-dataframes) for `dataframe` mode usage.

### Writing data

Input data can be:
1) A string properly formatted in InfluxDB's line protocol
2) A dictionary containing the following keys: `measurement`, `time`, `tags`, `fields`
3) A Pandas DataFrame with a DatetimeIndex
4) An iterable of one of the above

Input data in formats 2-4 are parsed into the 
[line protocol](https://docs.influxdata.com/influxdb/v1.3/write_protocols/line_protocol_reference/) 
before being written to InfluxDB. 
All parsing functionality is located at [`serialization.py`](aioinflux/serialization.py).
Beware that serialization is not highly optimized (PRs are welcome!) and may become a bottleneck depending 
on your application.


The `write` method returns `True` when successful and raises an `InfluxDBError` otherwise.  


#### Writing dictionary-like objects

Aioinflux accepts any dictionary-like object (mapping) as input. However, that dictionary must 
be properly formated and contain the following keys:

1) **`measurement`**: Optional. Must be a string-like object. If ommited, must be specified when 
  calling `AsyncInfluxDBClient.write` by passing a `measurement` argument.
1) **`time`**: Optional. The value can be `datetime.datetime`, date-like string 
  (e.g., `2017-01-01`, `2009-11-10T23:00:00Z`) or anything else that can be parsed by Pandas 
  [`Timestamp`](https://pandas.pydata.org/pandas-docs/stable/timeseries.html) class.
1) **`tags`**: Optional. This must contain another mapping of field names and values.
  Both tag keys and values should be strings.  
1) **`fields`**: Mandatory. This must contain another mapping of field names and values.
  Field keys should be strings. Field values can be `float`, `int`, `str`, or `bool` or any equivalent type. 

Any fields other then the above will be ignored when writing data to InfluxDB.


A typical dictionary-like point would look something like the following:

```python
{'time': '2009-11-10T23:00:00Z',
'measurement': 'cpu_load_short',
'tags': {'host': 'server01', 'region': 'us-west'},
'fields': {'value1': 0.64, 'value2': True, 'value3': 10}}
```


#### Writing DataFrames

Aioinflux also accepts Pandas dataframes as input. The only requirements for the dataframe is
that the **index must be of type `DatetimeIndex`**. Also, any column whose `dtype` is `object` will be 
converted to a string representation.

A typical DataFrame input should look something like the following:

```text
                                       LUY       BEM       AJW tag
2017-06-24 08:45:17.929097+00:00  2.545409  5.173134  5.532397   B
2017-06-24 10:15:17.929097+00:00 -0.306673 -1.132941 -2.130625   E
2017-06-24 11:45:17.929097+00:00  0.894738 -0.561979 -1.487940   B
2017-06-24 13:15:17.929097+00:00 -1.799512 -1.722805 -2.308823   D
2017-06-24 14:45:17.929097+00:00  0.390137 -0.016709 -0.667895   E
```

The measurement name must be specified with the `measurement` argument when calling `AsyncInfluxDBClient.write`.
Additional tags can also be passed using arbitrary keyword arguments.
 
**Example:**

```python
client = AsyncInfluxDBClient(db='testdb', mode='blocking')
client.write(df, measurement='prices', tag_columns=['tag'], asset_class='equities')
```

In the example above, `df` is the DataFrame we are trying to write to InfluxDB and 
`measurement` is the measurement we are writing to. 

`tag_columns` is in an optional iterable 
telling which of the dataframe columns should be parsed as tag values. If `tag_columns` is 
not explitly passed, all columns in the dataframe will be treated as InfluxDB field values.

Any other keyword arguments passed to `AsyncInfluxDBClient.write` are treated as extra tags 
which will be attached to the data being written to InfluxDB. Any string which is a valid 
[InfluxDB identifier](https://docs.influxdata.com/influxdb/v1.3/query_language/spec/#identifiers) 
and valid [Python identifier](https://docs.python.org/3/reference/lexical_analysis.html#identifiers) 
can be used as an extra tag key (with the exception of they strings `data`, `measurement` and `tag_columns`).

See `AsyncInfluxDBClient.write` docstring for details.


### Querying data

Querying data is as simple as passing an InfluxDB query string to `AsyncInfluxDBClient.write`:

```python
client.query('SELECT myfield FROM mymeasurement')
```

The result (in `blocking` and `async` modes) is a dictionary containing the raw JSON data 
returned by the InfluxDB 
[HTTP API](https://docs.influxdata.com/influxdb/v1.3/guides/querying_data/#querying-data-using-the-http-api):

```python
{'results': [{'series': [{'columns': ['time', 'Price', 'Volume'],
     'name': 'mymeasurement',
     'values': [[1491963424224703000, 5783, 100],
      [1491963424375146000, 5783, 200],
      [1491963428374895000, 5783, 100],
      [1491963429645478000, 5783, 1100],
      [1491963429655289000, 5783, 100],
      [1491963437084443000, 5783, 100],
      [1491963442274656000, 5783, 900],
      [1491963442274657000, 5782, 5500],
      [1491963442274658000, 5781, 3200],
      [1491963442314710000, 5782, 100]]}],
   'statement_id': 0}]}
```


#### Retrieving DataFrames

```text
                                  Price  Volume
2017-04-12 02:17:04.224703+00:00   5783     100
2017-04-12 02:17:04.375146+00:00   5783     200
2017-04-12 02:17:08.374895+00:00   5783     100
2017-04-12 02:17:09.645478+00:00   5783    1100
2017-04-12 02:17:09.655289+00:00   5783     100
2017-04-12 02:17:17.084443+00:00   5783     100
2017-04-12 02:17:22.274656+00:00   5783     900
2017-04-12 02:17:22.274657+00:00   5782    5500
2017-04-12 02:17:22.274658+00:00   5781    3200
2017-04-12 02:17:22.314710+00:00   5782     100
```

#### Chunked responses

TODO

#### Convenience functions

Aioinflux provides some wrappers around ``AsyncInfluxDBClient.query`` in order to provide
convenient access to commonly used query patterns. Appropriate named arguments must be 
passed (e.g.: `db`, `measurement`, etc).

Examples:

```python
client.create_database(db='foo')
client.drop_measurement(measurement='bar')
client.show_users()
```
For more complex queries, pass a raw query to ``AsyncInfluxDBClient.query``.

Please refer to the [source](aioinflux/client.py#L199) for argument information and to 
InfluxDB [documentation](https://docs.influxdata.com/influxdb/v1.3/query_language/spec/#queries)
for further query-related information.


### Other functionality

#### Authentication

TODO

#### Database selection

After the instantiation of the `AsyncInfluxDBClient` object, database can be switched 
by changing the `db` attribute:

```python
client = AsyncInfluxDBClient(db='db1')  # instatiate client
client.db = 'db2'  # switch database
```  

Beaware that differently from some NoSQL databases (such as MongoDB), InfluxDB requires that 
a databases is explicitly created (by using the 
[`CREATE DATABASE`](https://docs.influxdata.com/influxdb/v1.3/query_language/database_management/#create-database)
 query) before doing any operations on it.


#### Debugging

TODO


## Implementation

Since InfluxDB exposes all its functionality through an 
[HTTP API](https://docs.influxdata.com/influxdb/v1.3/tools/api/), 
`AsyncInfluxDBClient` tries to be nothing more than a thin and dry wrapper around that API.

The InfluxDB HTTP API exposes exactly three endpoints/functions: `ping`, `write` and `query`. 

`AsyncInfluxDBClient` merely wraps these three functions and provides some parsing functionality for generating 
line protocol data (when writing) and parsing JSON responses (when querying). 

Additionally, [partials](https://en.wikipedia.org/wiki/Partial_application) are used in 
order to provide convenient access to commonly used query patterns. 
See the [Convenience functions](#convenience-functions) section for details.


## Contributing

To contribute, fork the repository on GitHub, make your changes and submit a pull request.  
Aioinflux is not a mature project yet, so just simply raising issues is also greatly appreciated :)


CHANGELOG.md
# Changelog

## [0.9.0] - 2019-07-11

### Added
- Add support for custom path to InfluxDB (#24)
- Add support for `Decimal` serialization (812c1a8, 100d931)
- Add chunk count on chunked response debugging message (b9e85ad)

### Changed
- Refactor `rm_none` option implementation (5735b51, 13062ed, 89bae37)
- Make enum typevars more strict (f177212) 


## [0.8.0] - 2019-05-10

This is version is backwards compatible with v0.7.x

### Added
- Add dataframe support for chunked queries (e3c2a0b)


## [0.7.1] - 2019-04-11

This is version is backwards compatible with v0.7.0

### Fixed
- Don't cache error responses (be7b87c)

### Docs
- Minor wording changes

### Internal
- Minor internal changes 



## [0.7.0] - 2019-03-22

This is version is mostly backwards compatible with v0.6.x 
(w/ the exception of query patterns functionality)

## Added
- Redis-based caching functionality. See the 
  [docs](https://aioinflux.readthedocs.io/en/latest/usage.html#caching-query-results) for details.
- Timeout functionality (#21 by @SuminAndrew)

### Changed
- Move `ClientSession` creation logic outside `__init__`. 
It is now easier to used advanced ``aiohttp.ClientSession`` options.
See the [docs](https://aioinflux.readthedocs.io/en/latest/usage.html#other-aiohttp-functionality) for details. 

### Removed
- Query patterns functionality

### Internal
- Refactor test suite
- Various other internal changes 


## [0.6.1] - 2019-02-01

This is version is backwards compatible with v0.6.0

### Fixed
- Type annotation error in Python 3.6 (febfe47)
- Suppress `The object should be created from async function` warning from aiohttp 3.5 (da950e9) 


## [0.6.0] - 2019-02-01

### Added
- Support serializing NaN integers in pandas 0.24+ 
  (See [blog post](https://pandas-dev.github.io/pandas-blog/pandas-extension-arrays.html)) (1c55217)
- Support for using `namedtuple` with `iterpoints` (bd93c53)

### Changed
- **[BREAKING]** Changed signature of `parser` argument of `iterpoints`  
  from `(x, meta)` to `(*x, meta)` (bd93c53)

### Removed
- **[BREAKING]** Removed iterable mode and `InfluxDBResult` / `InfluxDBChunkedResult`.
  Use `iterpoints` instead. (592c5ed)
- Deprecated `set_query_pattern` (1d36b07)

### Docs
- Various improvements (8c6cbd3, ce46596, b7db169, ba3edae)


## [0.5.1] - 2019-01-21

This is version is backwards compatible with v0.5.0

### Fixed
- Fix type annotations
- Fix internal API inconsistencies

### Docs
- Complete API section
- Add proper Sphinx links
- Update/fix various sections


## [0.5.0] - 2019-01-17

### Changed
- [BREAKING] Removed `DataPoint` functionality in favor of simpler and more
 flexible `@lineprotocol` decorator. See the 
  [docs](https://aioinflux.readthedocs.io/en/latest/usage.html#writing-user-defined-class-objects) for details.

### Docs
- Added detailed `@lineprotocol` usage 
 
 
## [0.4.1] - 2018-11-22

### Fixed
- Fixed bug when doing multi-statement queries when using `dataframe` mode

### Docs
- Added note regarding handling of multi-statement/multi-series queries when using `dataframe` mode

 
## [0.4.0] - 2018-10-22

### Added
- Added ability to write `datapoint` objects. See the 
  [docs](https://aioinflux.readthedocs.io/en/latest/usage.html#writing-datapoint-objects) for details.
- Added `bytes` output format. This is to facilitate the addition of a caching layer on top of InfluxDB. (cb4e3d1)

### Changed
- Change `write` method signature to match the `/write` endpoint docs 
    - Allow writing to non-default retention policy (#14)
    - (`precision` is not fully implemented yet)
- Renamed `raw` output format to `json`. Most users should be unaffected by this. (cb4e3d1)

### Fixed
- Improved docs

### Internal
- Refactored serialization/parsing functionality into a subpackage
- Fix test warnings (2e42d50)

 

## [0.3.4] - 2018-09-03
- Fixed `output='dataframe'` parsing bug (#15)
- Removed tag column -> categorical dtype conversion functionality
- Moved documentation to Read The Docs
- Added two query patterns (671013b)
- Added this CHANGELOG


## [0.3.3] - 2018-06-23
- Python 3.7 support
- Sphinx-based documentation hosted at Read the Docs
- Minor dataframe serialization debugging (364190fa)

## [0.3.2] - 2018-05-03
- Fix parsing bug for string ending in a backslash (db8846e)
- Add InfluxDBWriteError exception class (d8d0a01)
- Make InfluxDBClient.db attribute optional (039e088)

## [0.3.1] - 2018-04-29
- Fix bug where timezone-unaware datetime input was assumed to be in local time (#11 / a8c81b7)
- Minor improvement in dataframe parsing (1e33b92)

## [0.3.0] - 2018-04-24
### Highlights:

- Drop Pandas/Numpy requirement (#9)
- Improved iteration support (816a722)
- - Implement tag/key value caching (9a65787)
- Improve dataframe serialization
  - Speed improvements (ddc9ecc)
  - Memory usage improvements (a2b58bd)
  - Disable concatenating of dataframes of the same measurement when grouping by tag (331a0c9)
  - Queries now return tag columns with `pd.Categorical` dtype (efdea98)
  - Writes now automatically identify `pd.Categorical` dtype columns as tag columns (ddc9ecc)

### API changes:

- `mode` attribute was "split" into `mode` and `output`.
 Default behavior remains the same (async / raw).
- Iteration is now made easier through the `iterable` mode 
 and `InfluxDBResult` and `InfluxDBChunkedResult` classes


## [0.2.0] - 2018-03-20

### Highlights

- Documentation is now complete
- Improved iteration support (via `iter_resp`) (cfffbf5)
- Allow users to add custom query patterns
- Add support for positional arguments in query patterns
- Reimplement `__del__` (40d0a69 / #7)
- Improve/debug dataframe parsing (7beeb53 / 96d78a4)
- Improve write error message (7972946) (by @miracle2k)

### API changes:
- Rename `AsyncInfluxDBClient` to `InfluxDBClient` (54d98c9)
- Change return format of chunked responses (related: cfffbf5 / #6)
- Make some `__init__` arguments keyword-only (5d2edf6)


## [0.1.2] - 2018-02-28


aioinflux/client.py
import asyncio
import json
import logging
import warnings
from functools import wraps
from typing import TypeVar, Union, AnyStr, Mapping, Iterable, Optional, AsyncGenerator

import aiohttp

from . import serialization
from .compat import *

if pd:
    PointType = TypeVar('PointType', Mapping, dict, bytes, pd.DataFrame)
    ResultType = TypeVar('ResultType', dict, bytes, pd.DataFrame)
else:
    PointType = TypeVar('PointType', Mapping, dict, bytes)
    ResultType = TypeVar('ResultType', dict, bytes)

# Aioinflux uses logging mainly for debugging purposes.
# Please attach your own handlers if you need logging.
logger = logging.getLogger('aioinflux')


def runner(coro):
    """Function execution decorator."""

    @wraps(coro)
    def inner(self, *args, **kwargs):
        if self.mode == 'async':
            return coro(self, *args, **kwargs)
        return self._loop.run_until_complete(coro(self, *args, **kwargs))

    return inner


class InfluxDBError(Exception):
    """Raised when an server-side error occurs"""
    pass


class InfluxDBWriteError(InfluxDBError):
    """Raised when a server-side writing error occurs"""

    def __init__(self, resp):
        self.status = resp.status
        self.headers = resp.headers
        self.reason = resp.reason
        super().__init__(f'Error writing data ({self.status} - {self.reason}): '
                         f'{self.headers.get("X-Influxdb-Error", "")}')


class InfluxDBClient:
    def __init__(
        self,
        host: str = 'localhost',
        port: int = 8086,
        path: str = '/',
        mode: str = 'async',
        output: str = 'json',
        db: Optional[str] = None,
        database: Optional[str] = None,
        ssl: bool = False,
        *,
        unix_socket: Optional[str] = None,
        username: Optional[str] = None,
        password: Optional[str] = None,
        timeout: Optional[Union[aiohttp.ClientTimeout, float]] = None,
        loop: Optional[asyncio.AbstractEventLoop] = None,
        redis_opts: Optional[dict] = None,
        cache_expiry: int = 86400,
        **kwargs
    ):
        """
        :class:`~aioinflux.client.InfluxDBClient`  holds information necessary
        to interact with InfluxDB.
        It is async by default, but can also be used as a sync/blocking client.
        When querying, responses are returned as parsed JSON by default,
        but can also be wrapped in easily iterable
        wrapper object or be parsed to Pandas DataFrames.
        The three main public methods are the three endpoints of the InfluxDB API, namely:

        1. :meth:`~.InfluxDBClient.ping`
        2. :meth:`~.InfluxDBClient.write`
        3. :meth:`~.InfluxDBClient.query`

        See each of the above methods documentation for further usage details.

        See also: https://docs.influxdata.com/influxdb/latest/tools/api/

        :param host: Hostname to connect to InfluxDB.
        :param port: Port to connect to InfluxDB.
        :param path: Path to connect to InfluxDB.
        :param mode: Mode in which client should run. Available options:

           - ``async``: Default mode. Each query/request to the backend will
           - ``blocking``: Behaves in sync/blocking fashion,
             similar to the official InfluxDB-Python client.

        :param output: Output format of the response received from InfluxDB.

           - ``json``: Default format.
             Returns parsed JSON as received from InfluxDB.
           - ``dataframe``: Parses results into :py:class`pandas.DataFrame`.
             Not compatible with chunked responses.

        :param db: Default database to be used by the client.
        :param ssl: If https should be used.
        :param unix_socket: Path to the InfluxDB Unix domain socket.
        :param username: Username to use to connect to InfluxDB.
        :param password: User password.
        :param timeout: Timeout in seconds or :class:`aiohttp.ClientTimeout` object
        :param database: Default database to be used by the client.
            This field is for argument consistency with the official InfluxDB Python client.
        :param loop: Asyncio event loop.
        :param redis_opts: Dict fo keyword arguments for :func:`aioredis.create_redis`
        :param cache_expiry: Expiry time (in seconds) for cached data
        :param kwargs: Additional kwargs for :class:`aiohttp.ClientSession`
        """
        self._loop = loop or asyncio.get_event_loop()
        self._session: aiohttp.ClientSession = None
        self._redis: aioredis.Redis = None
        self._mode = None
        self._output = None
        self._db = None
        self.ssl = ssl
        self.host = host
        self.port = port
        self.path = path
        self.mode = mode
        self.output = output
        self.db = database or db

        # ClientSession configuration
        if username:
            kwargs.update(auth=aiohttp.BasicAuth(username, password))
        if unix_socket:
            kwargs.update(connector=aiohttp.UnixConnector(unix_socket, loop=self._loop))
        if timeout:
            if isinstance(timeout, aiohttp.ClientTimeout):
                kwargs.update(timeout=timeout)
            else:
                kwargs.update(timeout=aiohttp.ClientTimeout(total=timeout))
        self.opts = kwargs

        # Cache configuration
        self.redis_opts = redis_opts
        self.cache_expiry = cache_expiry

    async def create_session(self, **kwargs):
        """Creates an :class:`aiohttp.ClientSession`

        Override this or call it with ``kwargs`` to use other :mod:`aiohttp`
        functionality not covered by :class:`~.InfluxDBClient.__init__`
        """
        self.opts.update(kwargs)
        self._session = aiohttp.ClientSession(**self.opts, loop=self._loop)
        if self.redis_opts:
            if aioredis:
                self._redis = await aioredis.create_redis(**self.redis_opts,
                                                          loop=self._loop)
            else:
                warnings.warn(no_redis_warning)

    @property
    def url(self):
        protocol = "https" if self.ssl else "http"
        return f"{protocol}://{self.host}:{self.port}{self.path}{{endpoint}}"

    @property
    def mode(self):
        return self._mode

    @property
    def output(self):
        return self._output

    @property
    def db(self):
        return self._db

    @mode.setter
    def mode(self, mode):
        if mode not in ('async', 'blocking'):
            raise ValueError('Invalid running mode')
        self._mode = mode

    @output.setter
    def output(self, output):
        if pd is None and output == 'dataframe':
            raise ValueError(no_pandas_warning)
        if output not in ('json', 'dataframe'):
            raise ValueError('Invalid output format')
        self._output = output

    @db.setter
    def db(self, db):
        self._db = db
        if not db:
            warnings.warn(f'No default databases is set. '
                          f'Database must be specified when querying/writing.')

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    def __del__(self):
        if not self._loop.is_closed() and self._session:
            asyncio.ensure_future(self._session.close(), loop=self._loop)

    def __repr__(self):
        items = [f'{k}={v}' for k, v in vars(self).items() if not k.startswith('_')]
        items.append(f'mode={self.mode}')
        return f'{type(self).__name__}({", ".join(items)})'

    @runner
    async def close(self):
        if self._session:
            await self._session.close()
            self._session = None
        if self._redis:
            self._redis.close()

    @runner
    async def ping(self) -> dict:
        """Pings InfluxDB

         Returns a dictionary containing the headers of the response from ``influxd``.
         """
        if not self._session:
            await self.create_session()
        async with self._session.get(self.url.format(endpoint='ping')) as resp:
            logger.debug(f'{resp.status}: {resp.reason}')
            return dict(resp.headers.items())

    @runner
    async def write(
        self,
        data: Union[PointType, Iterable[PointType]],
        measurement: Optional[str] = None,
        db: Optional[str] = None,
        precision: Optional[str] = None,
        rp: Optional[str] = None,
        tag_columns: Optional[Iterable] = None,
        **extra_tags,
    ) -> bool:
        """Writes data to InfluxDB.
        Input can be:

        1. A mapping (e.g. ``dict``) containing the keys:
           ``measurement``, ``time``, ``tags``, ``fields``
        2. A Pandas :class:`~pandas.DataFrame` with a :class:`~pandas.DatetimeIndex`
        3. A user defined class decorated w/
            :func:`~aioinflux.serialization.usertype.lineprotocol`
        4. A string (``str`` or ``bytes``) properly formatted in InfluxDB's line protocol
        5. An iterable of one of the above

        Input data in formats 1-3 are parsed to the line protocol before being
        written to InfluxDB.
        See the `InfluxDB docs <https://docs.influxdata.com/influxdb/latest/
        write_protocols/line_protocol_reference/>`_ for more details.

        :param data: Input data (see description above).
        :param measurement: Measurement name. Mandatory when when writing DataFrames only.
            When writing dictionary-like data, this field is treated as the default value
            for points that do not contain a `measurement` field.
        :param db: Database to be written to. Defaults to `self.db`.
        :param precision: Sets the precision for the supplied Unix time values.
            Ignored if input timestamp data is of non-integer type.
            Valid values: ``{'ns', 'u', 'µ', 'ms', 's', 'm', 'h'}``
        :param rp: Sets the target retention policy for the write.
            If unspecified, data is written to the default retention policy.
        :param tag_columns: Columns to be treated as tags
            (used when writing DataFrames only)
        :param extra_tags: Additional tags to be added to all points passed.
        :return: Returns ``True`` if insert is successful.
            Raises :py:class:`ValueError` otherwise.
        """
        if not self._session:
            await self.create_session()
        if precision is not None:
            # FIXME: Implement. Related issue: aioinflux/pull/13
            raise NotImplementedError("'precision' parameter is not supported yet")
        data = serialization.serialize(data, measurement, tag_columns, **extra_tags)
        params = {'db': db or self.db}
        if rp:
            params['rp'] = rp
        url = self.url.format(endpoint='write')
        async with self._session.post(url, params=params, data=data) as resp:
            if resp.status == 204:
                return True
            raise InfluxDBWriteError(resp)

    @runner
    async def query(
        self,
        q: AnyStr,
        *,
        epoch: str = 'ns',
        chunked: bool = False,
        chunk_size: Optional[int] = None,
        db: Optional[str] = None,
        use_cache: bool = False,
    ) -> Union[AsyncGenerator[ResultType, None], ResultType]:
        """Sends a query to InfluxDB.
        Please refer to the InfluxDB documentation for all the possible queries:
        https://docs.influxdata.com/influxdb/latest/query_language/

        :param q: Raw query string
        :param db: Database to be queried. Defaults to `self.db`.
        :param epoch: Precision level of response timestamps.
            Valid values: ``{'ns', 'u', 'µ', 'ms', 's', 'm', 'h'}``.
        :param chunked: If ``True``, makes InfluxDB return results in streamed batches
            rather than as a single response.
            Returns an AsyncGenerator which yields responses
            in the same format as non-chunked queries.
        :param chunk_size: Max number of points for each chunk. By default, InfluxDB chunks
            responses by series or by every 10,000 points, whichever occurs first.
        :param use_cache:
        :return: Response in the format specified by the combination of
           :attr:`.InfluxDBClient.output` and ``chunked``
        """

        async def _chunked_generator(url, data, dataframe):
            async with self._session.post(url, data=data) as resp:
                logger.debug(f'{resp.status} (CHUNKED): {q!r}')
                # Hack to avoid aiohttp raising ValueError('Line is too long')
                # The number 128 is arbitrary (may be too large/small).
                resp.content._high_water *= 128
                chunk_count = 0
                async for chunk in resp.content:
                    chunk = json.loads(chunk)
                    self._check_error(chunk)
                    chunk_count += 1
                    logger.debug(f'Yielding chunk #{chunk_count:03d}')
                    if dataframe:
                        yield serialization.dataframe.parse(chunk)
                    else:
                        yield chunk

        if not self._session:
            await self.create_session()

        # InfluxDB documentation is wrong regarding `/query` parameters
        # See https://github.com/influxdata/docs.influxdata.com/issues/1807
        if not isinstance(chunked, bool):
            raise ValueError("'chunked' must be a boolean")
        data = dict(q=q, db=db or self.db, chunked=str(chunked).lower(), epoch=epoch)
        if chunked and chunk_size:
            data['chunk_size'] = chunk_size

        url = self.url.format(endpoint='query')
        if chunked:
            if use_cache:
                raise ValueError("Can't use cache w/ chunked queries")
            elif self.mode != 'async':
                raise ValueError("Can't use 'chunked' with non-async mode")
            else:
                return _chunked_generator(url, data, self.output == 'dataframe')

        key = f'aioinflux:{q}'
        if use_cache and self._redis and await self._redis.exists(key):
            logger.debug(f'Cache HIT: {q}')
            data = lz4.decompress(await self._redis.get(key))
        else:
            async with self._session.post(url, data=data) as resp:
                data = await resp.read()
                if use_cache and self._redis:
                    logger.debug(f'Cache MISS ({resp.status}): {q}')
                    if resp.status == 200:
                        await self._redis.set(key, lz4.compress(data))
                        await self._redis.expire(key, self.cache_expiry)
                else:
                    logger.debug(f'{resp.status}: {q}')

        data = json.loads(data)
        self._check_error(data)
        if self.output == 'json':
            return data
        elif self.output == 'dataframe':
            return serialization.dataframe.parse(data)
        else:
            raise ValueError('Invalid output format')

    @staticmethod
    def _check_error(response):
        """Checks for JSON error messages and raises Python exception"""
        if 'error' in response:
            raise InfluxDBError(response['error'])
        elif 'results' in response:
            for statement in response['results']:
                if 'error' in statement:
                    msg = '{d[error]} (statement {d[statement_id]})'
                    raise InfluxDBError(msg.format(d=statement))

    # InfluxQL - Data management
    # --------------------------

    def create_database(self, db=None):
        db = db or self.db
        return self.query(f'CREATE DATABASE "{db}"')

    def drop_database(self, db=None):
        db = db or self.db
        return self.query(f'DROP DATABASE "{db}"')

    def drop_measurement(self, measurement):
        return self.query(f'DROP MEASUREMENT "{measurement}"')

    # InfluxQL - Schema exploration
    # -----------------------------

    def show_databases(self):
        return self.query("SHOW DATABASES")

    def show_measurements(self):
        return self.query("SHOW MEASUREMENTS")

    def show_users(self):
        return self.query("SHOW USERS")

    def show_series(self, measurement=None):
        if measurement:
            return self.query(f"SHOW SERIES FROM {measurement}")
        return self.query("SHOW SERIES")

    def show_tag_keys(self, measurement=None):
        if measurement:
            return self.query(f"SHOW TAG KEYS FROM {measurement}")
        return self.query("SHOW TAG KEYS")

    def show_field_keys(self, measurement=None):
        if measurement:
            return self.query(f"SHOW FIELD KEYS FROM {measurement}")
        return self.query("SHOW FIELD KEYS")

    def show_tag_values(self, key, measurement=None):
        if measurement:
            return self.query(f'SHOW TAG VALUES FROM "{measurement}" WITH key = "{key}"')
        return self.query(f'SHOW TAG VALUES WITH key = "{key}"')

    def show_retention_policies(self):
        return self.query("SHOW RETENTION POLICIES")

    # InfluxQL - Other
    # ----------------

    def show_continuous_queries(self):
        return self.query("SHOW CONTINUOUS QUERIES")
